File: app_py/routers/form_fields.py
import re
from fastapi import APIRouter, Depends, Request, Form, HTTPException
from fastapi.responses import RedirectResponse, HTMLResponse, JSONResponse
from sqlalchemy import func
from sqlalchemy.orm import Session, joinedload

from app_py.database import get_db, db_execute_raw, table_exists, get_column_type
from app_py.dependencies import require_permission
from app_py.models import CmsUser, Form as FormModel, FormField
from app_py.templating import templates, get_context
from app_py.utils import url_for

router = APIRouter()

_FIELD_NAME_RE = re.compile(r"^[a-z][a-z0-9_]*$")
_CONTROL_TYPES = ("input", "textarea", "select", "radio", "checkbox", "date", "datetime", "number", "file", "editor", "relation")


def _add_column_to_table(table: str, field: FormField):
    type_map = {
        "number": "BIGINT",
        "relation": "BIGINT",
        "date": "DATE",
        "datetime": "DATETIME",
        "textarea": "TEXT",
        "editor": "TEXT",
        "file": "VARCHAR(500)",
        "radio": "VARCHAR(255)",
        "checkbox": "VARCHAR(255)",
        "select": "VARCHAR(255)",
    }
    col_type = type_map.get(field.form_control, "VARCHAR(255)")
    nullable = "NOT NULL" if field.is_required else "NULL"
    db_execute_raw(f"ALTER TABLE `{table}` ADD COLUMN `{field.field_name}` {col_type} {nullable}")


def _drop_column_from_table(table: str, column: str):
    db_execute_raw(f"ALTER TABLE `{table}` DROP COLUMN `{column}`")


def _modify_column_nullable(table: str, column: str, nullable: bool):
    """修改列的 NULL/NOT NULL 约束"""
    col_type = get_column_type(table, column)
    null_clause = "NOT NULL" if nullable else "NULL"
    db_execute_raw(f"ALTER TABLE `{table}` MODIFY COLUMN `{column}` {col_type} {null_clause}")


@router.get("/{form_id}", response_class=HTMLResponse)
async def index(
    request: Request,
    form_id: int,
    current_user: CmsUser = Depends(require_permission("forms", "read")),
    db: Session = Depends(get_db),
):
    form = db.query(FormModel).options(joinedload(FormModel.fields)).filter(FormModel.id == form_id).first()
    if not form:
        raise HTTPException(404)
    from app_py.dependencies import get_menu_form_groups
    menu = get_menu_form_groups(db, current_user)
    return templates.TemplateResponse(
        "form-fields/index.html",
        get_context(request, cms_user=current_user, db=db, menu_form_groups=menu, form=form),
    )


@router.post("/")
async def store(
    request: Request,
    form_id: int = Form(...),
    field_name: str = Form(...),
    label: str = Form(...),
    form_control: str = Form("input"),
    options: str = Form(""),
    attributes: str = Form(""),
    sort_order: int = Form(0),
    is_required: bool = Form(False),
    is_list_visible: bool = Form(True),
    current_user: CmsUser = Depends(require_permission("forms", "update")),
    db: Session = Depends(get_db),
):
    form = db.query(FormModel).filter(FormModel.id == form_id).first()
    if not form:
        raise HTTPException(404)
    field_name = field_name.lower().strip()
    if not _FIELD_NAME_RE.match(field_name):
        if "application/json" in request.headers.get("accept", ""):
            raise HTTPException(400, detail={"code": 1, "msg": "字段名只能包含小写字母、数字和下划线"})
        return RedirectResponse(url=url_for("form_fields_index", form_id=form_id), status_code=302)
    if form_control not in _CONTROL_TYPES:
        if "application/json" in request.headers.get("accept", ""):
            raise HTTPException(400, detail={"code": 1, "msg": "无效的表单控件类型"})
        return RedirectResponse(url=url_for("form_fields_index", form_id=form_id), status_code=302)
    if db.query(FormField).filter(FormField.form_id == form.id, FormField.field_name == field_name).first():
        if "application/json" in request.headers.get("accept", ""):
            raise HTTPException(400, detail={"code": 1, "msg": "该字段已存在"})
        return RedirectResponse(url=url_for("form_fields_index", form_id=form_id), status_code=302)
    max_order = db.query(func.max(FormField.sort_order)).filter(FormField.form_id == form.id).scalar() or 0
    sort_order = sort_order if sort_order > 0 else int(max_order) + 1
    field = FormField(
        form_id=form.id,
        field_name=field_name,
        label=label,
        form_control=form_control,
        options=options or None,
        attributes=attributes or None,
        sort_order=sort_order,
        is_required=bool(is_required),
        is_list_visible=bool(is_list_visible),
    )
    db.add(field)
    db.commit()
    db.refresh(field)
    _add_column_to_table(form.table_name, field)
    if "application/json" in request.headers.get("accept", ""):
        return JSONResponse({"code": 0, "msg": "添加成功", "data": {"id": field.id}})
    return RedirectResponse(url=url_for("form_fields_index", form_id=form_id), status_code=302)


@router.put("/field/{field_id}")
async def update(
    request: Request,
    field_id: int,
    label: str = Form(...),
    form_control: str = Form("input"),
    options: str = Form(""),
    attributes: str = Form(""),
    sort_order: int = Form(0),
    is_required: bool = Form(False),
    is_list_visible: bool = Form(True),
    current_user: CmsUser = Depends(require_permission("forms", "update")),
    db: Session = Depends(get_db),
):
    field = db.query(FormField).filter(FormField.id == field_id).first()
    if not field:
        raise HTTPException(404)
    if form_control not in _CONTROL_TYPES:
        if "application/json" in request.headers.get("accept", ""):
            raise HTTPException(400, detail={"code": 1, "msg": "无效的表单控件类型"})
        return RedirectResponse(url=url_for("form_fields_index", form_id=field.form_id), status_code=302)
    old_required = field.is_required
    field.label = label
    field.form_control = form_control
    field.options = options or None
    field.attributes = attributes or None
    field.sort_order = sort_order
    field.is_required = bool(is_required)
    field.is_list_visible = bool(is_list_visible)
    db.commit()
    if old_required != field.is_required:
        try:
            _modify_column_nullable(field.form.table_name, field.field_name, field.is_required)
        except Exception as e:
            return JSONResponse({"code": 1, "msg": f"数据库列约束同步失败: {e}"}, status_code=500)
    return JSONResponse({"code": 0, "msg": "更新成功"})


@router.delete("/field/{field_id}")
async def destroy(
    request: Request,
    field_id: int,
    current_user: CmsUser = Depends(require_permission("forms", "delete")),
    db: Session = Depends(get_db),
):
    field = db.query(FormField).filter(FormField.id == field_id).first()
    if not field:
        raise HTTPException(404)
    form = field.form
    _drop_column_from_table(form.table_name, field.field_name)
    db.delete(field)
    db.commit()
    return JSONResponse({"code": 0, "msg": "删除成功"})

File: internal/handlers/form_fields.go
package handlers

import (
	"encoding/json"
	"fmt"
	"html/template"
	"laycms/internal/models"
	"laycms/internal/rawsql"
	"laycms/internal/tmpl"
	"strings"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

var controlTypes = map[string]struct{}{
	"input": {}, "textarea": {}, "select": {}, "radio": {}, "checkbox": {}, "date": {}, "datetime": {},
	"number": {}, "file": {}, "editor": {}, "relation": {},
}

func formFieldsIndex(c *gin.Context) {
	u := CurrentUser(c)
	fid := atoi(c.Param("form_id"), 0)
	var f models.Form
	if err := gormDB().Preload("Fields", func(db *gorm.DB) *gorm.DB {
		return db.Order("sort_order ASC")
	}).First(&f, fid).Error; err != nil {
		c.Status(404)
		return
	}
	page := tmpl.BuildPage(c, gormDB(), u, "字段配置 - "+f.Name, "")
	var rows []map[string]any
	for _, ff := range f.Fields {
		m := map[string]any{
			"id": ff.ID, "field_name": ff.FieldName, "label": ff.Label,
			"form_control": ff.FormControl, "sort_order": ff.SortOrder,
			"is_required": ff.IsRequired, "is_list_visible": ff.IsListVisible,
		}
		if ff.Options != nil {
			m["options"] = *ff.Options
		}
		rows = append(rows, m)
	}
	b, _ := json.Marshal(rows)
	type frag struct {
		*tmpl.Page
		Form       models.Form
		FieldsJSON template.JS
	}
	_ = tmpl.RenderFragmentWith(c, "form_fields", &frag{page, f, template.JS(b)}, page)
}

func formFieldsStore(c *gin.Context) {
	fid := atoi(c.PostForm("form_id"), 0)
	var f models.Form
	if err := gormDB().First(&f, fid).Error; err != nil {
		c.Status(404)
		return
	}
	fieldName := strings.ToLower(strings.TrimSpace(c.PostForm("field_name")))
	if !rawsql.SafeIdent(fieldName) {
		if tmpl.WantsJSON(c) {
			c.JSON(400, gin.H{"code": 1, "msg": "字段名只能包含小写字母、数字和下划线"})
			return
		}
		c.Redirect(302, "/form-fields/"+c.PostForm("form_id"))
		return
	}
	ctrl := c.DefaultPostForm("form_control", "input")
	if _, ok := controlTypes[ctrl]; !ok {
		if tmpl.WantsJSON(c) {
			c.JSON(400, gin.H{"code": 1, "msg": "无效的表单控件类型"})
			return
		}
		c.Redirect(302, "/form-fields/"+c.PostForm("form_id"))
		return
	}
	var cnt int64
	gormDB().Model(&models.FormField{}).Where("form_id = ? AND field_name = ?", f.ID, fieldName).Count(&cnt)
	if cnt > 0 {
		if tmpl.WantsJSON(c) {
			c.JSON(400, gin.H{"code": 1, "msg": "该字段已存在"})
			return
		}
		c.Redirect(302, "/form-fields/"+c.PostForm("form_id"))
		return
	}
	var maxOrder int16
	gormDB().Raw("SELECT IFNULL(MAX(sort_order),0) FROM cms_form_fields WHERE form_id = ?", f.ID).Scan(&maxOrder)
	so := int16(atoi(c.PostForm("sort_order"), 0))
	if so <= 0 {
		so = maxOrder + 1
	}
	opts := c.PostForm("options")
	attr := c.PostForm("attributes")
	var op, at *string
	if strings.TrimSpace(opts) != "" {
		op = &opts
	}
	if strings.TrimSpace(attr) != "" {
		at = &attr
	}
	ff := models.FormField{
		FormID: f.ID, FieldName: fieldName, Label: c.PostForm("label"), FormControl: ctrl,
		Options: op, Attributes: at, SortOrder: so,
		IsRequired: c.PostForm("is_required") == "true" || c.PostForm("is_required") == "on" || c.PostForm("is_required") == "1",
		IsListVisible: c.PostForm("is_list_visible") != "false" && c.PostForm("is_list_visible") != "0",
	}
	gormDB().Create(&ff)
	addColumnForField(f.DataTable, &ff)
	if tmpl.WantsJSON(c) {
		c.JSON(200, gin.H{"code": 0, "msg": "添加成功", "data": gin.H{"id": ff.ID}})
		return
	}
	c.Redirect(302, tmpl.URLFor("form_fields_index", "form_id", fmt.Sprintf("%d", fid)))
}

func addColumnForField(table string, field *models.FormField) {
	typeMap := map[string]string{
		"number": "BIGINT", "relation": "BIGINT", "date": "DATE", "datetime": "DATETIME",
		"textarea": "TEXT", "editor": "TEXT", "file": "VARCHAR(500)", "radio": "VARCHAR(255)",
		"checkbox": "VARCHAR(255)", "select": "VARCHAR(255)",
	}
	ct := typeMap[field.FormControl]
	if ct == "" {
		ct = "VARCHAR(255)"
	}
	null := "NULL"
	if field.IsRequired {
		null = "NOT NULL"
	}
	_ = gormDB().Exec(fmt.Sprintf("ALTER TABLE `%s` ADD COLUMN `%s` %s %s", table, field.FieldName, ct, null)).Error
}

func formFieldsUpdate(c *gin.Context) {
	fid := atoi(c.Param("field_id"), 0)
	var field models.FormField
	if err := gormDB().Preload("Form").First(&field, fid).Error; err != nil {
		c.Status(404)
		return
	}
	ctrl := c.DefaultPostForm("form_control", "input")
	if _, ok := controlTypes[ctrl]; !ok {
		c.JSON(400, gin.H{"code": 1, "msg": "无效的表单控件类型"})
		return
	}
	oldReq := field.IsRequired
	field.Label = c.PostForm("label")
	field.FormControl = ctrl
	opts := c.PostForm("options")
	attr := c.PostForm("attributes")
	if strings.TrimSpace(opts) == "" {
		field.Options = nil
	} else {
		field.Options = &opts
	}
	if strings.TrimSpace(attr) == "" {
		field.Attributes = nil
	} else {
		field.Attributes = &attr
	}
	field.SortOrder = int16(atoi(c.PostForm("sort_order"), 0))
	field.IsRequired = c.PostForm("is_required") == "true" || c.PostForm("is_required") == "on" || c.PostForm("is_required") == "1"
	field.IsListVisible = c.PostForm("is_list_visible") != "false" && c.PostForm("is_list_visible") != "0"
	gormDB().Save(&field)
	if oldReq != field.IsRequired {
		modifyColumnNullable(field.Form.DataTable, field.FieldName, field.IsRequired)
	}
	c.JSON(200, gin.H{"code": 0, "msg": "更新成功"})
}

func modifyColumnNullable(table, column string, notNull bool) {
	ct := rawsql.ColumnType(gormDB(), table, column)
	null := "NULL"
	if notNull {
		null = "NOT NULL"
	}
	_ = gormDB().Exec(fmt.Sprintf("ALTER TABLE `%s` MODIFY COLUMN `%s` %s %s", table, column, ct, null)).Error
}

func formFieldsDestroy(c *gin.Context) {
	fid := atoi(c.Param("field_id"), 0)
	var field models.FormField
	if err := gormDB().Preload("Form").First(&field, fid).Error; err != nil {
		c.Status(404)
		return
	}
	_ = gormDB().Exec(fmt.Sprintf("ALTER TABLE `%s` DROP COLUMN `%s`", field.Form.DataTable, field.FieldName)).Error
	gormDB().Delete(&field)
	c.JSON(200, gin.H{"code": 0, "msg": "删除成功"})
}
